x-pack/osquerybeat/beater/osquerybeat.go

```diff
@@ -0,0 +1,384 @@
+
+	// Create temp directory for socket and possibly other things
+	// The unix domain socker path is limited to 108 chars and would
+	// not always be able to create in subdirectory
+	tmpdir, removeTmpDir, err := createTempDir()
+	if err != nil {
+		return err
+	}
+	defer removeTmpDir()
+
+	// Install osqueryd if needed
+	err = installOsquery(ctx, exedir)
+	if err != nil {
+		return err
+	}
+
+	// Start osqueryd child process
+	osd := osqueryd.OsqueryD{
+		RootDir:    exedir,
+		SocketPath: osqueryd.SocketPath(tmpdir),
+	}
+
+	// Connect publisher
+	bt.client, err = b.Publisher.Connect()
+	if err != nil {
+		return err
+	}
+
+	// Start osqueryd child process
+	osdCtx, osdCn := context.WithCancel(ctx)
+	defer osdCn()
+	osqDone, err := osd.Start(osdCtx)
+	if err != nil {
+		bt.log.Errorf("Failed to start osqueryd process: %v", err)
+		return err
+	}
+
+	// Connect to osqueryd socket. Replying on the client library retry logic that checks for the socket availability
+	bt.osqCli, err = osqueryd.NewClient(ctx, osd.SocketPath, osqueryd.DefaultTimeout)
+	if err != nil {
+		bt.log.Errorf("Failed to create osqueryd client: %v", err)
+		return err
+	}
+
+	// Watch input configuration updates
+	inputConfigCh := config.WatchInputs(ctx)
+
+	// Start queries execution scheduler
+	scheduler := NewScheduler(ctx, bt.query)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		scheduler.Run()
+	}()
+
+	// Load initial queries
+	streams, inputTypes := config.StreamsFromInputs(bt.config.Inputs)
+	sz := len(streams)
+	if sz > 0 {
+		scheduler.Load(streams)
+	}
+
+	// Agent actions handlers
+	var actionHandlers []*actionHandler
+	unregisterActionHandlers := func() {
+		// Unregister action handlers
+		if b.Manager != nil {
+			for _, ah := range actionHandlers {
+				b.Manager.UnregisterAction(ah)
+				ah.bt = nil
+			}
+		}
+	}
+
+	registerActionHandlers := func(itypes []string) {
+		unregisterActionHandlers()
+		// Register action handler
+		if b.Manager != nil {
+			for _, inType := range itypes {
+				ah := &actionHandler{
+					inputType: inType,
+					bt:        bt,
+				}
+				b.Manager.RegisterAction(ah)
+			}
+		}
+	}
+
+	setManagerPayload := func(itypes []string) {
+		if b.Manager != nil {
+			b.Manager.SetPayload(map[string]interface{}{
+				"input_types":     itypes,
+				"osquery_version": distro.OsquerydVersion(),
+			})
+		}
+	}
+
+LOOP:
+	for {
+		select {
+		case err = <-osqDone:
+			break LOOP // Exiting if osquery child process exited with error
+		case <-ctx.Done():
+			bt.log.Info("Wait osqueryd exit")
+			exitErr := <-osqDone
+			bt.log.Infof("Exited osqueryd process, error: %v", exitErr)
+			break LOOP
+		case inputConfigs := <-inputConfigCh:
+			streams, inputTypes = config.StreamsFromInputs(inputConfigs)
+			registerActionHandlers(inputTypes)
+			setManagerPayload(inputTypes)
+			scheduler.Load(streams)
+		}
+	}
+
+	// Unregister action handlers
+	unregisterActionHandlers()
+
+	// Wait for clean scheduler exit
+	wg.Wait()
```

Comment:
don’t like defending against wg.Wait. it’s usually a bug upstream that needs fixing, if it stalls.
otherwise you would be leaking go routines with no consequences.